New Editor Interface

kouloumos opened the pull request:
## Overview

This PR closes #280, it introduces a new Editor interface built using our own modified version of the [slate-transcript-editor](https://github.com/kouloumos/slate-transcript-editor). Additionally, it includes a significant refactoring of the related code to streamline logic and eliminate redundant code from previous iterations.

I have made an effort to break down the changes into granular commits to facilitate the review process. However, the commit 1d36b53469b2773916392dedf78ac652ee8b2399 could not be further divided. I hope the detailed commits elsewhere will assist reviewers in their evaluation.

## Details

The new Editor introduces a new data format (DPE), stored for each transcript in the [bitcointranscripts-metadata](https://github.com/bitcointranscripts/bitcointranscripts-metadata) repository and used as input for the `slate-transcript-editor`. When a user claims a transcript, we now create two branches: one for DPE and one for markdown. Similarly, when they submit, we open two Pull Requests, one for each repository.

## Testing the changes

You can test the changes by pointing your local instance to staging with the following `.env` variables
```.env
NEXT_PUBLIC_APP_TRANSCRIPTION_BASE_URL = "https://brilliant-growth-staging.up.railway.app"
NEXT_PUBLIC_APP_QUEUE_BASE_URL = "https://transcription-review-backend-staging.up.railway.app/api"
NEXT_PUBLIC_VERCEL_ENV = "staging"
```
Then, claim a transcript and test the new Editor. 

## Next Steps

- Support save upon submission.
- Improve support for transcripts with a single speaker.
- Improve support for "Audience" speaker attributions.

_Note: the fork of [slate-transcript-editor](https://github.com/kouloumos/slate-transcript-editor) will be moved under the bitcointranscripts org at a later time._

kouloumos commented on src/services/api/github/useGithub.ts:
```diff
@@ -0,0 +1,320 @@
+    try {
+        await githubApi.get(dpeUrl);
+    } catch (error) {
+        throw new Error(`The DPE file does not exist for this transcript`);
```
This is not the deepgram file. This is the DPE file which is **required** as it is the input to the editor. This is just a sanity check. The DPE file should be available for all the transcripts in the queue.

kouloumos replied:
This is where you should be looking: https://github.com/bitcointranscripts/bitcointranscripts-metadata/tree/staging

The DPE data for all the currently queued transcripts on staging are there. So you should be able to claim transcripts.

We only use the deepgram files when we are constructing the DPE files. So the deepgram files are not used anywhere in the review process.

Emmanuel-Develops replied:
I sent the wrong link my bad.

The transcript I tried claiming on staging (Chaincode Decoded: Bech32m) does not have the dpe file.

<img width="1308" alt="Screenshot 2024-07-12 at 1 30 55 PM" src="https://github.com/user-attachments/assets/55c54e11-cfbe-4204-9136-d2e09441e3f6">

Here: https://github.com/bitcointranscripts/bitcointranscripts-metadata/tree/staging/chaincode-labs/chaincode-podcast/chaincode-decoded-bech32m

and many others.

kouloumos replied:
> and many others.

This is probably the only transcript in the queue without a DPE file. Try another one.

Emmanuel-Develops replied:
This too.
https://github.com/bitcointranscripts/bitcointranscripts-metadata/tree/staging/bitcoinplusplus/layer-2/exploring-p2p-hashrate-markets-v2

I noticed the first 2 didn't work. Tried the third one and I could claim it

0tuedon commented:
So, I did check out the UI work on the Editor, good work but I did encounter some issues

1.  Does the work with the playback feature??
<img width="1440" alt="Screenshot 2024-07-12 at 14 04 02" src="https://github.com/user-attachments/assets/456057db-6f44-4238-a339-6ffc9ae27e22">
I encounter a situation where the audio didn't play.


Also took a while to get used to this editor. Also this doesn't support MD?

kouloumos commented:
> I encounter a situation where the audio didn't play. 

Please send a screen recording with the issue.

> Also this doesn't support MD?

No. The markdown is generated upon save from the DPE file.

Emmanuel-Develops reviewed (commented):
Awesome features!
Here are some observations I noticed.

In-Editor Observations
- There are situations where I'd like to change the timestamps, for example here's a snippet of a transcript
```
00:00:20   Aaron van Wirdum    Live from Utrecht, this is Bitcoin Explained. Hello Sjors.
00:00:25   Sjors Provoost.          What's up? Welcome back. Thank you.
```
and it should be this

```
00:00:20   Aaron van Wirdum    Live from Utrecht, this is Bitcoin Explained. Hello Sjors.
00:00:25   Sjors Provoost           What's up?
00:00:26   Aaron van Wirdum    Welcome back.
00:00:27   Sjors Provoost           Thank you.
```
~~There's no way to edit that on the editor for now. In this case it's a minor issue, but there's no way to resolve it than creating a commit outside the editor.~~
Apparently, using a `shift+enter` creates new timestamps. Attached is a screen recording.

https://github.com/user-attachments/assets/9301bf6f-9e13-499e-8aaf-5a166fbaf303

- Audio Playback issues.
Sometimes the audio refuses to play, probably the same issue @0tuedon noticed. Although if I leave it for a while it resumes. 
Could be bandwidth related (my internet speed is about 20mbps) probably redownloading everytime one skips rather than downloading to memory.

- minor stuck state on exit speaker selection.
If I exit the speaker selection by clicking on itself, the cancel and save buttons don't work.
Screen recording below

https://github.com/user-attachments/assets/ef3e3997-ba96-4cca-839f-96b2cf5b6b49

- Unable to create new lines: As mentioned earlier, `shift+enter` creates new timestamps, users are unable to create new lines in the editor. This could be useful for quotes in markdown i.e
> quote here

Without new lines, some markdown tags do break. 

- submit doesn't save the last changes made to the editor.

kouloumos commented:
Regarding @Emmanuel-Develops comments:

> Unable to create new lines: [...] This could be useful for quotes in markdown i.e

As I replied [here](https://github.com/bitcointranscripts/transcription-review-front-end/pull/297#issuecomment-2225589135) markdown is not supported. The markdown file is created upon save/submission.

An onboarding flow will be added later to better explain to users what's available and how to use the interface. This includes the way to add new timestamps including everything that is currently available when you hover over the "How Does this work?" text.

> submit doesn't save the last changes made to the editor.

This is a known limitation, I will include it in the description. This functionality will be added in subsequent PR.

> minor stuck state on exit speaker selection.

This seems to be because of the way the dropdown is implemented. I'll see what I can do to fix it.

Extheoisah commented on src/utils/transcript.ts:
```diff
@@ -1,8 +1,7 @@
 import config from "@/config/config.json";
 
-export const compareTranscriptBetweenSave = (data: SaveToGHData) => {
+export const compareTranscriptBetweenSave = (data: any) => {
```
why is the data type changed to `any`? I saw you removed the `SaveToGHData` type as well. Is there a reason for this? If extra data types were added due to the new data points, we should update the type

kouloumos replied:
You can see in the rest of the codebase that this type was removed because it was not needed after the refactoring because the logic now uses existing types.

In this occasion I just switched to `any` while developing and forgot to update. I'll change it.

Extheoisah replied:
Makes sense

Extheoisah reviewed (commented):
Quite a long PR. 
In general, the pr looks good. I left some comments and nits.
I haven't tested locally though.

kouloumos commented on next.config.js:
```diff
@@ -10,6 +13,16 @@ const nextConfig = {
+    config.resolve.alias['slate-transcript-editor'] = require.resolve('slate-transcript-editor/dist');
+    config.resolve.alias = {
+      ...config.resolve.alias,
+      react: path.resolve('./node_modules/react'),
```
This was needed while I was linking a local instance of the `slate-transcript-editor` for development purposes as it was picking the versions of react in that one. Just checked and those are not needed with the current setup which consumes the package from the GitHub repo. I assume that this will also be the case if/when we use a published package.

Good catch! I will remove it, and check if it still needed when developing locally.

kouloumos replied:
This is still needed when developing with a local slate-transcript-editor. Removing it would create issues for anyone trying to do that in the future, so I decided to leave it and add a comment there to explain why the code is there.

There might be a way to make those aliases obsolete, but I already spend too much time trying to do that, so I leave it as it is for now.

Extheoisah replied:
sounds good

kouloumos commented on next.config.js:
```diff
@@ -1,3 +1,6 @@
+const withTM = require("next-transpile-modules")(["slate-transcript-editor"]);
```
I'll check it out. Tagging @IgboPharaoh in case he has anything to add to this as he was the one that wrote this setup in `next.config.js`

kouloumos replied:
changed this to `transpilePackages`

kouloumos commented on src/components/modals/SelectSpeakerModal.tsx:
```diff
@@ -0,0 +1,97 @@
+    const { data: allMetadata } = useGetMetadata();
+
+
+    useLayoutEffect(() => {
```
Yes, there was an inconsistency where a previous version of selected speaker was shown in the modal selection, and because `useLayoutEffect` "fires before the browser repaints the screen" ([src](https://react.dev/reference/react/useLayoutEffect)), that issues was solved by using it.

I just read in the linked docs that it can hurt performance. I don't recall of putting a lot of effort to see if this issue was rooted somewhere else, thus avoiding `useLayoutEffect`. I'll check it out.

kouloumos replied:
Good catch, I improved the state handling logic and `useLayoutEffect` is not needed anymore.

kouloumos commented on src/components/modals/SubmitTranscriptModal.tsx:
```diff
@@ -1,151 +1,228 @@
+                <Text>
+                  Corrections and conversations about this transcript will
+                  shift to GitHub on your PR linked
+                  <span> </span>
```
This is previously written code that I just moved to a new component. I'll check if that's what `span` is for and replace with your suggestion if needed.

kouloumos commented on src/components/modals/SuggestModal.tsx:
```diff
@@ -77,42 +75,21 @@ export function SuggestModal({ handleClose, isOpen }: SuggestModalProps) {
-          });
-          resetAndCloseForm()
-        },
+    await suggestSource.mutateAsync({
```
You will find the callback inside the `useGitHub` hook.
I added all the callbacks inside of the hooks for better separation of concerns. This `onSuccess` is the only one that "survived" because it uses context from the component.

kouloumos commented on src/components/sideBarContentEdit/SelectField.tsx:
```diff
@@ -184,11 +185,17 @@ export const OnlySelectField = ({
   autoCompleteList,
   userCanAddToList,
   horizontal,
+  single
```
If you see all the changes in the commit (6c75a1e81c58592b56e33c86e10829cc0142eda6) that introduced this change, you will observe that this flag is not just to allow you to select one speaker, it also changes the way that the dropdown displays the selected speaker, that's why the flag is needed.

kouloumos replied:
I changed the implementation and the `single` flag is not needed. The behavior of the component is now derived from the `editedData` that are given to the `SelectField`.

kouloumos commented on src/pages/api/github/read.ts:
This route reads a file from a branch, so `get-file-from-branch` could be suitable.

That said, the rest of the routes are called `fork`, `pr`, `newBranch`, `save`, that's why `read` seemed more suitable to me.

If we are going to change this one, I think that it makes sense to also rename the rest of the routes.

Extheoisah replied:
In that case, let's leave it as it is

kouloumos commented on src/services/api/github/useGithub.ts:
```diff
@@ -0,0 +1,320 @@
+    await githubApi.post('/newBranch', {
+        upstreamRepo,
+        // TODO: needs better handling of base branch
+        baseBranch: process.env.NEXT_PUBLIC_VERCEL_ENV === "production" ? "master" : "staging",
```
That's what the TODO is about.

kouloumos commented on src/services/api/reviews/useReview.ts:
```diff
@@ -1,19 +1,110 @@
+      transcript: { metadata, dpe, url: reviewData.transcript.transcriptUrl },
+    };
+  } catch (error: any) {
+    throw new Error(
```
probably not, I'll verify.

kouloumos replied:
I verified that this is something that we are doing in multiple places in the codebase. It is needed to better handle error messages from http requests

kouloumos commented on src/components/modals/SelectSpeakerModal.tsx:
```diff
@@ -0,0 +1,97 @@
+                        onChange={(e) => setIsUpdateAllSpeakerInstances(e.target.checked)}
+                        mt={4}
+                    >
+                        Replace all occurrences of {oldSpeakerName}
```
All the string in this page (and in the app) are written in the same way. So I will leave it as it is.

kouloumos commented on src/components/modals/SubmitTranscriptModal.tsx:
```diff
@@ -1,151 +1,228 @@
+            <Text>Date of original presentation</Text>
+          </ListItem>
+          <ListItem>
+            <Text>Tags for the main topics of discussion</Text>
```
I did a grammar check on this and I don't think that your suggestion is correct. This is because "topics" is plural, indicating that there are multiple topics being tagged. Therefore, "main topics" should also be plural to agree with "Tags."

Extheoisah replied:
Isn't it just one topic or is it multiple topics? If the context here refers to multiple topics, then you're correct. You can discard my suggestion then.

kouloumos replied:
multiple topics

kouloumos commented on src/services/api/reviews/useReview.ts:
```diff
@@ -1,19 +1,110 @@
+const getMetadata = async (branchUrl: string): Promise<TranscriptMetadata> => {
+  const response = await axios.post(`/api/github/read`, { branchUrl });
+  const content = response.data.content;
+  const matches = /---\n([\s\S]+?)\n---/.exec(content);
```
done

kouloumos commented on src/components/modals/SelectSpeakerModal.tsx:
```diff
@@ -0,0 +1,97 @@
+    };
+
+    const updateSpeaker = (speakers: string[]) => {
+        setNewSpeakerName(speakers[0])
```
This was related with my comment [here](https://github.com/bitcointranscripts/transcription-review-front-end/pull/297#discussion_r1684121325). But I since changed the implementation and the component can work without extra flags for both an array of speakers and single speakers.

kouloumos commented:
I updated the PR with all of @Extheoisah suggestions and I also fixed the "minor stuck state on exit speaker selection" issue pointed by @Emmanuel-Develops.
